test/viper/ok/text.vpr.ok

```diff
@@ -0,0 +1,170 @@
+/* BEGIN PRELUDE */
+/* Array encoding */
+domain Array {
+  function $loc(a: Array, i: Int): Ref
+  function $size(a: Array): Int
+  function $loc_inv1(r: Ref): Array
+  function $loc_inv2(r: Ref): Int
+  axiom $all_diff_array { forall a: Array, i: Int :: {$loc(a, i)} $loc_inv1($loc(a, i)) == a && $loc_inv2($loc(a, i)) == i }
+  axiom $size_nonneg { forall a: Array :: $size(a) >= 0 }
+}
+define $array_acc(a, t, p) forall j: Int :: 0 <= j && j < $size(a) ==> acc($loc(a, j).t, p)
+define $array_untouched(a, t) forall j: Int :: 0 <= j && j < $size(a) ==> $loc(a, j).t == old($loc(a, j).t)
+define $array_init(a, t, x) forall i : Int :: 0 <= i && i < $size(a) ==> $loc(a, i).t == x
+/* Tuple encoding */
+domain Tuple {
+  function $prj(a: Tuple, i: Int): Ref
+  function $prj_inv1(r: Ref): Tuple
+  function $prj_inv2(r: Ref): Int
+  axiom $all_diff_tuple { forall a: Tuple, i: Int :: {$prj(a, i)} $prj_inv1($prj(a, i)) == a && $prj_inv2($prj(a, i)) == i }
+}
+/* Option encoding */
+adt Option[T] {
+  None()
+  Some(some$0: T)
+}
+/* Text encoding */
+function $concat(a: Int, b: Int): Int
+/* Typed references */
+field $int: Int
+field $bool: Bool
+field $ref: Ref
+field $array: Array
+field $tuple: Tuple
+field $option_int: Option[Int]
+field $option_bool: Option[Bool]
+field $option_array: Option[Array]
+field $option_tuple: Option[Tuple]
+/* END PRELUDE */
+
+field $text_8: Int
+field $text_6: Int
+field $text_4: Int
+field $text_2: Int
+field $text: Int
+define $Perm($Self) ((((((((true && acc(($Self).fld1,write)) && acc(($Self).fld2,write)) && 
+  acc(($Self).$text_8,write)) && acc(($Self).$text_6,write)) && acc(($Self).$text_4,write)) && 
+  acc(($Self).$text_2,write)) && acc(($Self).$text,write)))
+define $Inv($Self) (true)
+method __init__($Self: Ref)
+    
+    requires $Perm($Self)
+    ensures $Perm($Self)
+    ensures $Inv($Self)
+    { 
+      ($Self).fld1 := ($Self).$text_8;
+      ($Self).fld2 := ($Self).$text_8; 
+    }
+field fld1: Int
+field fld2: Int
+method concat_text($Self: Ref)
+    
+    requires $Perm($Self)
+    ensures $Perm($Self)
+    { var x: Int
+      var y: Int
+      var z: Int
+      x := ($Self).$text;
+      y := ($Self).$text_2;
+      z := $concat(x, y);
+      assert (z == $concat(($Self).$text, ($Self).$text_2));
```

Comment:
Why text literal is encoded as an actor field?
I think this could obfuscate permissions of actor and load the solver by unneeded work.

Have you considered simpler approach e.g. replace all literal by a number (in first approximation, even different numbers for occurrences of the same literal). Or use function without body `alloc_str(): Int`. 
Or even simpler: compile `let x: Text = "hello"` into `var x: Int` i.e. omit init expressions.

Reply:
> replace all literal by a number

Applied this.

> in first approximation, even different numbers for occurrences of the same literal

For example this won't work for the next assertion
```
assert:system "foo" == "foo"
```
because these two `"foo"` literals would be translated to different numbers.

> Or use function without body alloc_str(): Int.

The same reason as above but will result in a false negative
```
assert:system "foo" == "bar"
```
This assertion holds.

> Or even simpler: compile let x: Text = "hello" into var x: Int i.e. omit init expressions.

I wouldn't say that this is simpler. In my current implementation, I translate string literals directly. This approach wouldn't handle things like
```
let x: Text = "Hello" # "world";
```
We need to translate this statement into A-normal form.

The second reason this is harder is because uninitialized `var x: Int` could be produced only by emitting `decl` and `exp` translation returns only the `exp` object. That means that we should handle assignment statements, not the literals themself.

Moreover, I suspect that this translation won't work for the same reasons as for different literal IDs.

Reply:
>The same reason as above but will result in a false negative
assert:system "foo" == "bar"
This assertion holds.

You are right. This is not working

Reply:
>We need to translate this statement into A-normal form.
We can just assume that. I think when A-normal form will be implemented it might not be hard to add such case as well.

Reply:
I think solution with a global map that bind string literals to unique numbers is fine. But I would move it out of `ctxt` because now it contains only immutable structures and after changes also will contain mutable map. How functional update works in that case in Ocaml?

I would make this map global as for `fresh_id`

Reply:
>How functional update works in that case in Ocaml?

I see that Ocaml is copying table by reference. So it is working, but maybe global map is better? :thinking: 
Probably it will cause some inconvenience when we will translate several units, but I think for that we have to have explicit global (inter-unit) translator context and a separately unit-local translator context (that we have now).

Reply:
> it contains only immutable structures 
Okay actually also mutable references to immutable structures. Probably they are used in a same manner as you did. I am not familiar with that part of the code.

Reply:
Moved this `Hashtbl` to the global scope

Reply:
Thx, I think now changes is quite simple.